### ISSUE TITLE
Rename symbol: No result

### ISSUE THREAD

**Issue by chr7** [microsoft/pylance-release#1360]
I have a similar problem as reported in #441.

## Environment data

Pylance version: 2021.5.3
VSC Version: 1.56.2 (user setup)
Commit: 054a9295330880ed74ceaedda236253b4f39a335
Date: 2021-05-12T17:13:13.157Z
Electron: 12.0.4
Chrome: 89.0.4389.114
Node.js: 14.16.0
V8: 8.9.255.24-electron.0
OS: Windows_NT x64 10.0.17134
Python version 3.8.8 in a virtual pipenv environment, version 2020.11.15

## Expected behaviour

Selecting a variable, pressing F2 (or mouse right click and selecting 'Rename Symbol') and changing the name of the variable should rename the variable within the scope of the function.

## Actual behaviour

`No result.`

## Logs

```
XXX
```

## Code Snippet / Additional information

I used 'Open Folder...' and opened a file within one of the subfolders. The file with several functions already exists for a long time. Now I have the need to add a class. To start with, I copied the body of an already existing function to a new method of that class. 

### Obervations

1. Trying to rename a specifc variable within that class method fails ('No result') but renaming the same variable in the function where I copied the body from, works. The (simple) class looks correct, no errors are shown.
2. Renaming fails not for all variables within the class method, for example: 
```python
class Foo:
    def check(self, val):
        a = val
```
Renaming fails for 'val' but works for 'a'.

**Comment by chr7** [comment]
I found a way to reproduce it not using my code but creating an example.

I created a new file in one of my subfolders and entered the following code:
```
class FooClass:

    def check(self, value):
        a = val
        return a

```

Then I selected 'val', pressed F2 and entered 'value' (to get 'a = value') but instead of being renamed, I get the famous "No result" popup. Renaming 'a' (e.g. to 'b') works.

**Comment by jakebailey** [comment]
That doesn't seem like the same thing, though; `val` isn't actually declared anywhere, it's an undefined variable and we squiggle that. Is that the case with your original code?

**Comment by chr7** [comment]
That's excactly what I did with my code: I created a class and the method 'check', copied the (lengthy) body from another function and noticed that this body uses 'val' while I decided to name the parameter of the method 'value'. So I decided to stick with 'value' and rename 'val' to 'value' in the body of the method - and that's excatly what is not working then, I get the "No result" popup.

**Comment by erictraut** [comment]
Renames are done semantically, not based on simple text substitutions. After all, you wouldn't want "val" to be renamed unconditionally throughout your file. In this particular case, there is no semantic meaning for "val" because it isn't a defined symbol.

I'm not sure why one would use the "Rename Symbol" feature in this particular case since it's less work to simply rename the single instance of the identifier in the editor. But I agree that the current behavior is confusing, so perhaps we should change the "Rename Symbol" command to do a single-instance replacement if there's no declaration found.

**Comment by jakebailey** [comment]
> But I agree that the current behavior is confusing, so perhaps we should change the "Rename Symbol" command to do a single-instance replacement if there's no declaration found.

Yes, I agree, if the only issue here is renaming an undefined variable; IIRC this is what TS does.

**Comment by chr7** [comment]
Ok, then this is my misunderstanding of how 'Rename Symbol' works.

The example above is simple, so for sure it's easy just to rename the single occurance of 'val' to 'value' manually in the example.

But my real live example is more complex and 'val' is used on different places within the body of the method. I could use the "Search & Replace" functionality but 'Rename Symbol':
1. is more comfortable. Hitting F2 and getting the field to enter the new name right below the variable is very comfortable. Moreover, I don't have to highlight the complete name, cursor on it is just enough.
2. it renames a symbol within a scope (method here), so I don't have to care about other occurances within the file. And that's excactly what I want to do here, renaming this variable only in the body of this method.

That's why I acctually use "Search & Replace" rather seldom (I can't remember when I used it the last time).

**Comment by jakebailey** [comment]
You use `val` more than once, but the `val` is not defined anywhere? If `val` is defined and isn't being squiggled, then the code sample differs from the original code.

**Comment by jakebailey** [comment]
Or are you saying that you copy the code in, then want to rename the undefined variable in all uses?

**Comment by chr7** [comment]
Yup, refactoring ... copy the code from elsewhere and renaming the variable then.

**Comment by jakebailey** [comment]
That's certainly a little more tricky. Renaming one is easy (just the one word), but knowing which undefined variables uses to also rename is harder when our binding process doesn't identify them to be the same (it's similar to what is there at runtime). I'd have to see if TS goes this far for a comparison.

**Comment by heejaechang** [comment]
I dont think this is a case where rename symbol would support. but to reduce confusion, I will make pylance to tell users rename is not available in this case. like how typescript currently does.

**Comment by bschnurr** [comment]
This issue has been fixed in version 2022.3.3, which we've just released. You can find the changelog here: [CHANGELOG.md](https://github.com/microsoft/pylance-release/blob/main/CHANGELOG.md#202233-23-March-2022)
